In [1]:
import pandas as pd

In [2]:
df=pd.read_csv(r"C:\Users\jayan\OneDrive\Desktop\To transfer\Internship\my-work-flyrank-main\my-work-flyrank\data\raw\content_refresh_anonymized.csv")

# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

In [3]:
df[["impressions_90d", "clicks_90d", "sessions_90d", "days_since_last_update"]]

,impressions_90d,clicks_90d,sessions_90d,days_since_last_update
0,3803,29,17,20
1,15320,7,9,25
2,12581,11,11,20
3,11751,58,78,22
4,19140,24,145,14
...,...,...,...,...
29995,1,0,3,20
29996,761,3,3,20
29997,6336,12,44,20
29998,154763,340,463,22


In [4]:
df.columns.tolist()

['content_id',
 'client_id',
 'search_volume',
 'competition',
 'competition_level',
 'cpc',
 'content_type',
 'main_intent',
 'word_count',
 'char_count',
 'provider_used',
 'model_used',
 'impressions_90d',
 'clicks_90d',
 'pageviews_90d',
 'sessions_90d',
 'users_90d',
 'engaged_sessions_90d',
 'ai_sessions_90d',
 'scroll_events_90d',
 'days_with_impressions',
 'days_with_sessions',
 'impressions_last_30d',
 'clicks_last_30d',
 'sessions_last_30d',
 'impressions_prev_30d',
 'clicks_prev_30d',
 'sessions_prev_30d',
 'content_age_days',
 'age_tier',
 'age_tier_order',
 'days_since_last_update',
 'freshness_tier',
 'word_count_tier',
 'char_count_tier',
 'ctr',
 'avg_position',
 'engagement_rate',
 'scroll_rate',
 'ai_traffic_pct',
 'impression_tier',
 'position_tier',
 'trend_direction',
 'trend_pct']

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

In [14]:
df["content_id"].is_unique

True

Here in this dataset one row represent one page and metrices related to it.
The 90d columns cover a 90-day window ending at the day when the details were recorded, but since there's no daye column, this window is relative rather than tied to a specific calendar date.
Also above we have confirmed that column content_is is unique and no datetime column exist in our dataset.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

Feature bucket - `impressions_90d`, `clicks_90d`, `ctr`, `engagement_rate`, `days_since_last_update`

Label - Empty

Context - `content_id`, `client_id`

Excluded bucket - `provider_used` , `ai_sessions_90d`, `ai_traffic_pct`, `trend_direction`, `trend_pct`

In [16]:
pd.set_option('display.max_columns', None)
df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,provider_used,model_used,impressions_90d,clicks_90d,pageviews_90d,sessions_90d,users_90d,engaged_sessions_90d,ai_sessions_90d,scroll_events_90d,days_with_impressions,days_with_sessions,impressions_last_30d,clicks_last_30d,sessions_last_30d,impressions_prev_30d,clicks_prev_30d,sessions_prev_30d,content_age_days,age_tier,age_tier_order,days_since_last_update,freshness_tier,word_count_tier,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,NaN,gemini-2.5-flash,3803,29,22,17,16,1,0,1,88,13,578,2,2,987,13,9,187,181-365,5,20,0-30,2000-3500,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,NaN,gemini-3-flash-preview,15320,7,10,9,9,0,0,1,88,9,2501,2,3,5915,1,2,445,365+,6,25,0-30,2000-3500,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,NaN,gemini-2.5-flash,12581,11,14,11,11,0,0,4,88,11,2382,1,1,6089,3,3,141,91-180,4,20,0-30,3500+,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,NaN,NaN,11751,58,87,78,75,1,0,3,88,51,3626,22,35,4206,17,26,463,365+,6,22,0-30,NaN,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,NaN,gemini-3-flash-preview,19140,24,177,145,144,0,0,43,88,33,4211,10,14,6452,2,9,263,181-365,5,14,0-30,2000-3500,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [19]:
print(df["content_id"].size)
df.isnull().sum()

30000


content_id                    0
client_id                     0
search_volume              2468
competition                2468
competition_level          2610
cpc                        2468
content_type                  0
main_intent                2374
word_count                 7699
char_count                 7699
provider_used             21438
model_used                 5733
impressions_90d               0
clicks_90d                    0
pageviews_90d                 0
sessions_90d                  0
users_90d                     0
engaged_sessions_90d          0
ai_sessions_90d               0
scroll_events_90d             0
days_with_impressions         0
days_with_sessions            0
impressions_last_30d          0
clicks_last_30d               0
sessions_last_30d             0
impressions_prev_30d          0
clicks_prev_30d               0
sessions_prev_30d             0
content_age_days              0
age_tier                      0
age_tier_order                0
days_sin

In [33]:
print(df["trend_direction"].value_counts())
print(df["trend_pct"].value_counts())
print(df["ai_sessions_90d"].value_counts())
print(df["ai_traffic_pct"].value_counts())
print(df["competition"].value_counts())

trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64
trend_pct
-100.0     1395
 0.0        443
-50.0       281
-66.7       172
 100.0      145
           ... 
 1079.4       1
 75.3         1
 109.4        1
 67.7         1
 577.6        1
Name: count, Length: 2712, dtype: int64
ai_sessions_90d
0     28070
1       700
2       585
3       165
4       153
5        67
6        61
7        40
8        37
9        26
10       17
11       17
12       13
13        7
14        5
18        4
20        4
15        4
19        3
26        3
16        3
25        2
21        2
38        1
41        1
40        1
24        1
60        1
46        1
17        1
64        1
23        1
59        1
35        1
39        1
Name: count, dtype: int64
ai_traffic_pct
0.00      28070
50.00        48
100.00       40
33.33        33
14.29        31
          ...  
3.91          1
2.11          1
6.58          1
0.11          1
134.62        1

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [34]:
df.groupby("content_type")["search_volume"].apply(lambda x: x.isna().mean())

content_type
comparison article    0.000000
feedly article        1.000000
keyword article       0.013673
Name: search_volume, dtype: float64

Clearly from above we can infer content_id is unique, there is no datetime data type in our data. From overall missingness counts we can clearly see that ctr and engagement_rate have no null values also by search_volume and content_type groupby we have proved structural missingness.

Additionally `provider_used` shows 21438 missing values, confirming it should be excluded due to high number of missing values and `ai_sessions_90d` shows that 28070 rows have value 0 so it should also be excluded due to being extremely being sparse

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

This data can never help me in comparison of different quarters because of missing datetime column.

My data type is richer for some types only. It is richer for comparison article whereas poorer for feedly article.

Never tells us about the reason for the behaviour of the pages this is because of lack of article text.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.